lib/Net/Async/Blockchain/BTC.pm

```diff
@@ -123,9 +125,18 @@ sub subscribe {
     $subscription = $subscription_dictionary{$subscription};
 
     die "Invalid or not implemented subscription" unless $subscription && $self->can($subscription);
+    my ($zmq_client_source, $zmq_client_socket) = $self->new_zmq_client->subscribe($subscription);
+
+    my $error_handler = sub {
+        my $error = shift;
+        $self->source->fail($error);
+        zmq_close($zmq_client_socket);
+    };
+
+## Please see file lib/Net/Async/Blockchain/BTC.pm.ERR
     Future->needs_all(
-        $self->new_zmq_client->subscribe($subscription)->map(async sub { await $self->$subscription(shift) })->ordered_futures->completed(),
-        $self->recursive_search());
+        $zmq_client_source->map(async sub { await $self->$subscription(shift) })->ordered_futures->completed(),
+        $self->recursive_search())->on_fail($error_handler)->retain;
```

Comment:
just to nit pick, shouldn't it be `on_fail(sub {...} )` ?

Reply:
the `$error_handler` is an anonymous subroutine

Reply:
ah ok. got it